libkbfs/chat_rpc.go

```diff
@@ -296,3 +309,180 @@ func (c *ChatRPC) ReadChannel(
 	}
 	return messages, nextPage, nil
 }
+
+// RegisterForMessages implements the Chat interface.
+func (c *ChatRPC) RegisterForMessages(
+	convID chat1.ConversationID, cb ChatChannelNewMessageCB) {
+	str := convID.String()
+	c.convLock.Lock()
+	defer c.convLock.Unlock()
+	c.convCBs[str] = append(c.convCBs[str], cb)
+}
+
+// We only register for the kbfs-edits type of notification in
+// keybase_daemon_rpc, so all the other methods below besides
+// `NewChatKBFSFileEditActivity` should never be called.
+var _ chat1.NotifyChatInterface = (*ChatRPC)(nil)
+
+// NewChatActivity implements the chat1.NotifyChatInterface for
+// ChatRPC.
+func (c *ChatRPC) NewChatActivity(
+	_ context.Context, _ chat1.NewChatActivityArg) error {
+	return nil
+}
+
+// NewChatKBFSFileEditActivity implements the
+// chat1.NotifyChatInterface for ChatRPC.
+func (c *ChatRPC) NewChatKBFSFileEditActivity(
+	ctx context.Context, arg chat1.NewChatKBFSFileEditActivityArg) error {
+	activityType, err := arg.Activity.ActivityType()
+	if err != nil {
+		return err
+	}
+	switch activityType {
+	case chat1.ChatActivityType_NEW_CONVERSATION:
+		// If we learn about a new conversation for a given TLF,
+		// attempt to route it to the TLF.
+		info := arg.Activity.NewConversation()
+		tlfType := tlf.Private
+		if info.Conv.Visibility == keybase1.TLFVisibility_PUBLIC {
+			tlfType = tlf.Public
+		} else if info.Conv.MembersType == chat1.ConversationMembersType_TEAM {
+			tlfType = tlf.SingleTeam
+		}
+		tlfHandle, err := GetHandleFromFolderNameAndType(
+			ctx, c.config.KBPKI(), c.config.MDOps(), info.Conv.Name, tlfType)
+		if err != nil {
+			return err
+		}
+		c.config.KBFSOps().NewNotificationChannel(
+			ctx, tlfHandle, info.ConvID, info.Conv.Channel)
```

Comment:
Why does the registration happen inside `NewNotificationChannel` instead of here? Seems like we're introducing a back-and-forth dependency.

Reply:
Because there might not yet be an FBO/edit history object for the corresponding TLF yet, and so we won't have anything to register.  And we don't want to create one if one isn't being used yet.  (We'll only create edit history objects for the top N chats in our inbox.)

This call is supposed to be a fire-and-forget type thing; any registration resulting from it should probably happen in a new goroutine.

Reply:
Understood, thanks.